src/plugins/intel_cpu/src/nodes/multinomial.cpp

```diff
@@ -0,0 +1,200 @@
+        num_samples_prc = InferenceEngine::Precision::I32;
+    }
+
+    auto out_prc = getOriginalOutputPrecisionAtPort(OUTPUT_PORT);
+    if (!one_of(out_prc, InferenceEngine::Precision::I64, InferenceEngine::Precision::I32)) {
+        out_prc = InferenceEngine::Precision::I32;
+    }
+
+    addSupportedPrimDesc({{LayoutType::nspc, probs_prc}, {LayoutType::nspc, num_samples_prc}},
+                         {{LayoutType::nspc, out_prc}},
+                         ref_any);
+}
+
+void Multinomial::prepareParams() {
+    auto probs_shape = getParentEdgeAt(PROBS_PORT)->getMemory().getStaticDims();
+    auto num_samples_shape = getParentEdgeAt(NUM_SAMPLES_PORT)->getMemory().getStaticDims();
+
+    if (probs_shape.size() != 1 && probs_shape.size() != 2) {
+        IE_THROW() << m_errorPrefix << "has incompatible 'probs' shape " << PartialShape(probs_shape)
+                   << ". Only 1D and 2D tensors are allowed.";
+    }
+
+    if (num_samples_shape.size() != 1) {
+        IE_THROW() << m_errorPrefix << "has incompatible 'num_samples' shape " << PartialShape(num_samples_shape)
+                   << ". Only scalar and 1D single element tensors are allowed.";
+    }
+
+    m_probs_1d = probs_shape.size() == 1;
+
+    m_input_elements_count = std::accumulate(probs_shape.begin(), probs_shape.end(), 1, std::multiplies<size_t>());
+    m_batches_count = probs_shape.size() == 2 ? probs_shape[0] : 1;
+    m_probs_count = probs_shape.size() == 2 ? probs_shape[1] : probs_shape[0];
+
+    m_cdf.reserve(m_input_elements_count);
+    m_input_vals.reserve(m_input_elements_count);
+    m_random_samples.reserve(m_input_elements_count);
+    m_max_per_batch.reserve(probs_shape.size() == 1 ? 1 : probs_shape[1]);
+}
+
+void Multinomial::execute(dnnl::stream strm) {
+    const auto* probs = reinterpret_cast<const float*>(getParentEdgeAt(PROBS_PORT)->getMemoryPtr()->getData());
+    const auto num_samples =
+        reinterpret_cast<const int*>(getParentEdgeAt(NUM_SAMPLES_PORT)->getMemoryPtr()->getData())[0];
+    auto* output = reinterpret_cast<int*>(getChildEdgeAt(OUTPUT_PORT)->getMemoryPtr()->getData());
+
+    // exp & cumsum
+    if (m_log_probs) {
+        parallel_for(m_input_elements_count, [&](size_t idx) {
+            m_input_vals[idx] = std::exp(probs[idx]);
+        });
+        m_cdf[0] = m_input_vals[0];
+        for (size_t idx = 1; idx < m_input_elements_count; ++idx) {
+            m_cdf[idx] = m_input_vals[idx] + m_cdf[idx - 1];
+        }
+    } else {
+        m_cdf[0] = probs[0];
+        for (size_t idx = 1; idx < m_input_elements_count; ++idx) {
+            m_cdf[idx] = probs[idx] + m_cdf[idx - 1];
+        }
+    }
+
+    // TODO RandomUniform - requires RandomUniform kernel https://github.com/openvinotoolkit/openvino/pull/20171
+    std::srand(m_op_seed);
+    parallel_for(m_input_elements_count, [&](size_t idx) {
+        m_random_samples[idx] = static_cast<double>(std::rand()) / static_cast<double>(RAND_MAX);
```

Comment:
Is there any reason to use double? Usually float is enough. Please try to use float after adding the tests.

Reply:
I tried to match the reference in case the probabilities are double values, but since f64 was removed then this can be float as well.